weed/filer/filer.go

```diff
@@ -240,6 +240,43 @@ func (f *Filer) CreateEntry(ctx context.Context, entry *Entry, o_excl bool, isFr
 	return nil
 }
 
+func (f *Filer) RenameEntry(ctx context.Context, oldEntry, newEntry *Entry, o_excl bool, isFromOtherCluster bool, signatures []int32, skipCreateParentDir bool, maxFilenameLength uint32) error {
+	if string(newEntry.FullPath) == "/" {
+		return nil
+	}
+
+	if newEntry.FullPath.IsLongerFileName(maxFilenameLength) {
+		return fmt.Errorf("entry name too long")
+	}
+
+	currentEntry, _ := f.FindEntry(ctx, newEntry.FullPath)
+	if currentEntry != nil && o_excl {
+		glog.V(3).Infof("EEXIST: entry %s already exists", newEntry.FullPath)
+		return fmt.Errorf("EEXIST: entry %s already exists", newEntry.FullPath)
+	}
+
+	if currentEntry == nil && !skipCreateParentDir {
+		dirParts := strings.Split(string(newEntry.FullPath), "/")
+		if err := f.ensureParentDirectoryEntry(ctx, newEntry, dirParts, len(dirParts)-1, isFromOtherCluster); err != nil {
+			return err
+		}
+	}
+
+	glog.V(4).Infof("renameEntry %s: old entry: %v", newEntry.FullPath, oldEntry.FullPath)
+	if err := f.renameEntry(ctx, oldEntry, newEntry, currentEntry != nil); err != nil {
+		glog.Errorf("rename entry %s: %v", newEntry.FullPath, err)
+		return fmt.Errorf("rename entry %s: %v", newEntry.FullPath, err)
+	}
+
+	f.NotifyUpdateEvent(ctx, oldEntry, newEntry, true, isFromOtherCluster, signatures)
```

Comment:
the `currentEntry` removal would be missing from the metadata events.

Reply:
if currentEntry exists, it is updated, not deleted.
for example: rename from a to b, and b exists, after renaming, a rename event of b will be generated, b is still there, so the rename event is also a update event of old b

Reply:
no. There are a, b, b_old 3 entries in this example. One event can only have info for 2 entries.

Reply:
yes, there are two events, rename: oldEntry: a(1234), newEntry b(1234), update: oldEntry: b(5678), newEntry: b(1234). only the 1st one is generated, do I need to generate the 2nd one?

Reply:
@chrislusf there are two events as I said above, a rename and an update, new entry are all b, and these two events happen exactly the same time. If we generate two events, what is the order?  a=>b and then b_old=>b or b_old=>b and then a=>b, both are meaningless, I think an event a=>b is enough.

In other words, what is a rename event? inode is the same between old entry and new entry, but name is different, it is a rename event.
If inode is different between old entry and new entry, but name is the same, what kind of event is it? it is not a update event since inode is changed. or an override event? or just a side effect of a rename event?